### PR TITLE
Add four regression tests

### PR DESCRIPTION
**Pull request by frewsxcv**

### CODE REVIEW

**Comment by alexcrichton** [comment]
@bors: r+ 0d326b50d490dbcac4a37f37a5791b3a23b3bc4f

Thanks!

**Comment by frewsxcv** [comment]
Woops! The expected error message for 23595-1.rs was incorrect; I just fixed it

**Comment by frewsxcv** [comment]
Cool, I had another typo. Hopefully it should be okay now, sorry about that

**Comment by alexcrichton** [comment]
@bors: r+ 28ff886

No worries!

**Comment by frewsxcv** [comment]
I'm not sure what I'm doing wrong with the expected error. If anyone has insight please let me know.

**Comment by frewsxcv** [comment]
I just learned from @bluss that when reporting multiple compile errors on a single line, one needs to incrementally add `^` for each line (`^`, `^^`, `^^^`, etc.). I just made that change so this should be good to go again

File: src/test/compile-fail/issue-23595-1.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// except according to those terms.
+
+use std::ops::{Index};
+
+trait Hierarchy {
+    type Value;
+    type ChildKey;
+    type Children = Index<Self::ChildKey, Output=Hierarchy>;
+    //~^ ERROR: the value of the associated type `ChildKey`
+    //~^^ ERROR: the value of the associated type `Children`
```
Comment (tamird) [inline src/test/compile-fail/issue-23595-1.rs]: you can also use `//~|` on this line i believe

Comment (tamird) [inline src/test/compile-fail/issue-23595-1.rs]: see https://github.com/rust-lang/rust/blob/master/src/test/compile-fail/array-not-vector.rs#L16

Comment (tamird) [inline src/test/compile-fail/issue-23595-1.rs]: docs: https://github.com/rust-lang/rust/blob/master/src/compiletest/errors.rs#L26

**Comment by alexcrichton** [comment]
@bors: r+ 7954096bc8b0519164f20a3b8178efb221fb86b0

but also r=me if you wanna tweak the syntax a bit

**Pull request merged**